index.js

```diff
@@ -97,84 +97,167 @@ const parseArgs = (
 
   let pos = 0;
   while (pos < argv.length) {
-    let arg = argv[pos];
-
-    if (StringPrototypeStartsWith(arg, '-')) {
-      if (arg === '-') {
-        // '-' commonly used to represent stdin/stdout, treat as positional
-        result.positionals = ArrayPrototypeConcat(result.positionals, '-');
-        ++pos;
-        continue;
-      } else if (arg === '--') {
-        // Everything after a bare '--' is considered a positional argument
-        // and is returned verbatim
-        result.positionals = ArrayPrototypeConcat(
-          result.positionals,
-          ArrayPrototypeSlice(argv, ++pos)
-        );
-        return result;
-      } else if (StringPrototypeCharAt(arg, 1) !== '-') {
-        // Look for shortcodes: -fXzy and expand them to -f -X -z -y:
-        if (arg.length > 2) {
-          for (let i = 2; i < arg.length; i++) {
-            const short = StringPrototypeCharAt(arg, i);
-            // Add 'i' to 'pos' such that short options are parsed in order
-            // of definition:
-            ArrayPrototypeSplice(argv, pos + (i - 1), 0, `-${short}`);
-          }
-        }
+    const arg = argv[pos];
+    const nextArg = argv[pos + 1];
 
-        arg = StringPrototypeCharAt(arg, 1); // short
-        if (options.short && options.short[arg])
-          arg = options.short[arg]; // now long!
-        // ToDo: later code tests for `=` in arg and wrong for shorts
-      } else {
-        arg = StringPrototypeSlice(arg, 2); // remove leading --
+    // Check if `arg` is an options terminator.
+    // Guideline 10 in https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html
+    if (arg === '--') {
+      // Everything after a bare '--' is considered a positional argument.
+      result.positionals = ArrayPrototypeConcat(
+        result.positionals,
+        ArrayPrototypeSlice(argv, pos + 1)
+      );
+      break; // Finished processing argv, leave while loop.
+    }
+
+    if (isLoneShortOption(arg)) {
+      // e.g. '-f'
+      const optionKey = getOptionKey(StringPrototypeCharAt(arg, 1), options);
+      let optionValue;
+      if (isExpectingValue(optionKey, options) && isOptionValue(nextArg)) {
+        // e.g. '-f' 'bar'
+        optionValue = nextArg;
+        pos++;
+      }
+      storeOptionValue(optionKey, optionValue, options, result);
+      pos++;
+      continue;
+    }
+
+    if (isShortOptionGroup(arg, options)) {
+      // Expand -fXzy to -f -X -z -y
+      const expanded = [];
+      for (let index = 1; index < arg.length; index++) {
+        ArrayPrototypePush(expanded, `-${StringPrototypeCharAt(arg, index)}`);
       }
+      // Replace group with expansion.
+      ArrayPrototypeSplice(argv, pos, 1, ...expanded);
```

Comment:
using array spread syntax can't be made robust; is there another way this can be done without using splice/spread?

Reply:
Replaced splice and spread with concat and slice.

For wry amusement, your earlier comment about splice: https://github.com/pkgjs/parseargs/pull/59/files#r800124055